File: Processor_Sim_Project/Old_C_Code(ScalarOnly)/Simulator/Load_Assembly.c
#include <Load_Assembly.h>

//load specified assembly program
//igrnores any lines beginning with '//' in the input program
void loadProgram(const char *filename, struct instruction *INSTR){
    //open file
    FILE *programFile = fopen(filename, "r");
    if(programFile == NULL){
        printf("ERROR - File '%s' not found.\n", filename);
        exit(1);
    }

    //get next line
    char line[LINE_LEN];
    while(fgets(line, LINE_LEN, programFile)){
        //temp instruction
        struct instruction newInstruction;
        //dont do anything with comment lines
        if(line[0] != '/' && line[1] != '/'){
            int wordCount = 0;
            int charCount = 0;
            char word[256];
            for(int j = 0; j < LINE_LEN; j++){
                //after each instruction part, convert to correct format
                if(line[j] == ' ' || line[j] == '\n'){
                    switch(wordCount){
                        case 0: newInstruction.opCode = opcodeSTRtoENUM(word);
                                break;
                        case 1: newInstruction.operand1 = operandSTRtoENUM(word);
                                break;
                        case 2: newInstruction.operand2 = operandSTRtoENUM(word);
                                break;
                        case 3: newInstruction.operand3 = operandSTRtoENUM(word);
                                break;
                    }
                    wordCount ++;
                    memset(word, 0, sizeof(word));
                    charCount = 0;
                    continue;
                }
                //move on if comma or end of line
                if(line[j] == ',') continue;
                if(line[j] == '\n') break;
                word[charCount] = line[j];
                charCount ++;
            }
            //ensures all instructions have a correct instruction opcode
            if(newInstruction.opCode == -1){
                printf("Program file '%s' OpCode read incorrectly for instruction %d. Please check.\n", filename, instructionCount);
                exit(1);
            }
            //store instruction
            INSTR[instructionCount] = newInstruction;
            instructionCount ++;
        }
    }
    fclose(programFile);
    instructionCount = 0;
}


//convert between instruction string read from file and enum version
int opcodeSTRtoENUM(char *opcode){
    if(strcmp(opcode, "HALT") == 0) return HALT;
    else if(strcmp(opcode, "ADD") == 0) return ADD;
    else if(strcmp(opcode, "ADDI") == 0) return ADDI;
    else if(strcmp(opcode, "SUB") == 0) return SUB;
    else if(strcmp(opcode, "SUBI") == 0) return SUBI;
    else if(strcmp(opcode, "MUL") == 0) return MUL;
    else if(strcmp(opcode, "MULI") == 0) return MULI;
    else if(strcmp(opcode, "DIV") == 0) return DIV;
    else if(strcmp(opcode, "DIVI") == 0) return DIVI;
    else if(strcmp(opcode, "LD") == 0) return LD;
    else if(strcmp(opcode, "LDC") == 0) return LDC;
    else if(strcmp(opcode, "STR") == 0) return STR;
    else if(strcmp(opcode, "STRC") == 0) return STRC;
    else if(strcmp(opcode, "CMP") == 0) return CMP;
    else if(strcmp(opcode, "JMP") == 0) return JMP;
    else if(strcmp(opcode, "BR") == 0) return BR;
    else if(strcmp(opcode, "BEQ") == 0) return BEQ;
    else if(strcmp(opcode, "BLT") == 0) return BLT;
    else return -1;
}

//converts operand into associated register enum index, or the directly to an int if an immediate operand
int operandSTRtoENUM(char *operand){
    if(strcmp(operand, "r0") == 0) return r0;
    else if(strcmp(operand, "r1") == 0) return r1;
    else if(strcmp(operand, "r2") == 0) return r2;
    else if(strcmp(operand, "r3") == 0) return r3;
    else if(strcmp(operand, "r4") == 0) return r4;
    else if(strcmp(operand, "r5") == 0) return r5;
    else if(strcmp(operand, "r6") == 0) return r6;
    else if(strcmp(operand, "r7") == 0) return r7;
    else if(strcmp(operand, "r8") == 0) return r8;
    else if(strcmp(operand, "r9") == 0) return r9;
    else if(strcmp(operand, "r11") == 0) return r11;
    else if(strcmp(operand, "r12") == 0) return r12;
    else if(strcmp(operand, "r13") == 0) return r13;
    else if(strcmp(operand, "r14") == 0) return r14;
    else if(strcmp(operand, "r15") == 0) return r15;
    else if(strcmp(operand, "r16") == 0) return r16;
    else if(strcmp(operand, "r17") == 0) return r17;
    else if(strcmp(operand, "r18") == 0) return r18;
    else if(strcmp(operand, "r19") == 0) return r19;
    else if(strcmp(operand, "r20") == 0) return r20;
    else if(strcmp(operand, "r21") == 0) return r21;
    else if(strcmp(operand, "r22") == 0) return r22;
    else if(strcmp(operand, "r23") == 0) return r23;
    else if(strcmp(operand, "r24") == 0) return r24;
    else if(strcmp(operand, "r25") == 0) return r25;
    else if(strcmp(operand, "r26") == 0) return r26;
    else if(strcmp(operand, "r27") == 0) return r27;
    else if(strcmp(operand, "r28") == 0) return r28;
    else if(strcmp(operand, "r29") == 0) return r29;
    else if(strcmp(operand, "r30") == 0) return r30;
    else if(strcmp(operand, "r31") == 0) return r31;
    //defualt being convert the string to int
    else return atoi(operand);
}

File: Processor_Sim_Project/Scalar_Sim_Submission/Python_Sim/Load_Assembly.py
import sys
from Instruction import Instruction

# Line length constant for reading assembly lines from text file
LINE_LEN = 256

# Instructions:
# HALT = 0, ADD = 1, ADDI = 2, SUB = 3, SUBI = 4, MUL = 5, MULI = 6, DIV = 7, DIVI = 8, LD = 9,
# LDC = 10, STR = 11, STRC = 12, CMP = 13, JMP = 14, BR = 15, BEQ = 16, BLT = 17


# Registers. int corresponding value ==> rx = int(x)    i.e. r13 = 13
# r0, r1, r2, r3, r4, r5, r6, r7, r8, r9, r10, r11, r12, r13, r14, r15, r16, 
# r17, r18, r19, r20, r21, r22, r23, r24, r25, r26, r27, r28, r29, r30, r31

# Load specified assembly program
# Igrnores any lines beginning with '//' in the input program
def loadProgram(filename, INSTR) :
    instructionCount = 0
    # Open file
    programFile = open(filename, 'r')
    if(programFile == None) :
        print("ERROR - File '%s' not found.\n", filename)
        sys.exit(1)

    while True :
        line = programFile.readline()
        # Stop the loop at end of file
        if not line :
            break

        words = line.split(' ')
        # Remove comments
        if '//' in words[0] :
            continue
        # Remove commas and line endings
        for i in range(1, len(words)) :
            words[i] = words[i][:-1]

        newOpcode = words[0]
        newOperands = ['0'] * 3
        for i in range(1, len(words)) :
            newOperands[i-1] = words[i]

        INSTR[instructionCount] = Instruction(opcodeSTRtoENUM(newOpcode), operandSTRtoENUM(newOperands[0]), operandSTRtoENUM(newOperands[1]), operandSTRtoENUM(newOperands[2]), 0)
        instructionCount += 1

    programFile.close()
        

# Convert between instruction string read from file and enum version
def opcodeSTRtoENUM(opcode) :
    if(opcode == "HALT") :
        return 0
    elif(opcode == "ADD") :
        return 1
    elif(opcode == "ADDI") :
        return 2
    elif(opcode == "SUB") :
        return 3
    elif(opcode == "SUBI") :
        return 4
    elif(opcode == "MUL") :
        return 5
    elif(opcode == "MULI") :
        return 6
    elif(opcode == "DIV") :
        return 7
    elif(opcode == "DIVI") :
        return 8
    elif(opcode == "LD") :
        return 9
    elif(opcode == "LDC") :
        return 10
    elif(opcode == "STR") :
        return 11
    elif(opcode == "STRC") :
        return 12
    elif(opcode == "CMP") :
        return 13
    elif(opcode == "JMP") :
        return 14
    elif(opcode == "BR") :
        return 15
    elif(opcode == "BEQ") :
        return 16
    elif(opcode == "BLT") :
        return 17
    return -1

# Converts operand into associated register enum index, or the directly to an int if an immediate operand
def operandSTRtoENUM(operand) :
    if(operand == "r0") :
        return 0
    elif(operand == "r1") :
        return 1
    elif(operand == "r2") :
        return 2
    elif(operand == "r3") :
        return 3
    elif(operand == "r4") :
        return 4
    elif(operand == "r5") :
        return 5
    elif(operand == "r6") :
        return 6
    elif(operand == "r7") :
        return 7
    elif(operand == "r8") :
        return 8
    elif(operand == "r9") :
        return 9
    elif(operand == "r10") :
        return 10
    elif(operand == "r11") :
        return 11
    elif(operand == "r12") :
        return 12
    elif(operand == "r13") :
        return 13
    elif(operand == "r14") :
        return 14
    elif(operand == "r15") :
        return 15
    elif(operand == "r16") :
        return 16
    elif(operand == "r17") :
        return 17
    elif(operand == "r18") :
        return 18
    elif(operand == "r19") :
        return 19
    elif(operand == "r20") :
        return 20
    elif(operand == "r21") :
        return 21
    elif(operand == "r22") :
        return 22
    elif(operand == "r23") :
        return 23
    elif(operand == "r24") :
        return 24
    elif(operand == "r25") :
        return 25
    elif(operand == "r26") :
        return 26
    elif(operand == "r27") :
        return 27
    elif(operand == "r28") :
        return 28
    elif(operand == "r29") :
        return 29
    elif(operand == "r30") :
        return 30
    elif(operand == "r31") :
        return 31

    # Defualt being convert the string to int
    return int(operand)